### PR TITLE
enhancement(elasticsearch sink): add support for data streams

### PR DESCRIPTION
**Pull request by spencergilbert**
Adds a configuration option to specify 'index' (default) or 'create' (for data streams)

Related: https://github.com/timberio/vector/issues/4939

Signed-off-by: Spencer Gilbert <spencer.gilbert@gmail.com>

### CODE REVIEW

**Comment by spencergilbert** [comment]
I'll look at the test failures later today

File: src/sinks/elasticsearch.rs [hunk 1]
```diff
@@ -94,6 +95,35 @@ impl ElasticSearchAuth {
...
+
+impl BulkAction {
+    pub fn as_str(&self) -> &'static str {
+        match *self {
+            BulkAction::Index => "index",
+            BulkAction::Create => "create",
+        }
+    }
+
+    pub fn as_path(&self) -> &'static str {
```
Comment (jszwedko) [inline src/sinks/elasticsearch.rs:119]: Maybe `as_json_pointer`. I initially thought this was going to be used as an HTTP path.

**Review by jszwedko — COMMENTED** [review]
This seems like a reasonable implementation to me!

We'll just want to update the docs and probably add a test as you noted in discord.

**Comment by jszwedko** [comment]
You can ignore the windows test failure, we are working on that one.

**Comment by spencergilbert** [comment]
@jszwedko somewhat confused by what's needed for the test failures - I might have misunderstood you comment in discord, but compiler doesn't like adding serde(default)

**Comment by jszwedko** [comment]
@spencergilbert Ah, I meant I think you need `#[serde(default)]` on https://github.com/timberio/vector/pull/5126/files#diff-767cdc952613316aab8665fa7c9ae09a5126bde478de907bcfa3e8010e6ec31eR64

**Comment by portante** [comment]
How does this code handle 409 conflicts when `create` is used and the object is already present?

**Comment by portante** [comment]
FWIW, we try to gracefully handle the Elasticsearch responses in [pyesbulk](https://github.com/distributed-system-analysis/py-es-bulk/blob/master/pyesbulk/__init__.py#L511).

**Comment by spencergilbert** [comment]
> How does this code handle 409 conflicts when `create` is used and the object is already present?

:thinking: I didn't add any specific handling so it would end up using the existing retry logic here - https://github.com/timberio/vector/blob/master/src/sinks/elasticsearch.rs#L289

**Comment by portante** [comment]
> https://github.com/timberio/vector/blob/master/src/sinks/elasticsearch.rs#L310

That looks good, just might be a bit noisy for duplicates, if the payload contains pre-generated IDs ('_id').

If the client is not generating their own IDs for the Elasticsearch records, then `create` not get 409s, and depending on how efficient Vector ends up being sending `_bulk` requests, it could generate many duplicates in certain cases.

Does the request code time out on long requests?  If so, that will be another problem when the Elasticsearch instance gets over loaded.  

The logic looks like it retries on server errors for any 500 error.  We only retry on 500, 503, and 504 with a backoff.

If the client is using `index`, or using `create` without generating their own IDs, then retries on 5xx errors can overwhelm and Elasticsearch instance.

**Comment by spencergilbert** [comment]
> > https://github.com/timberio/vector/blob/master/src/sinks/elasticsearch.rs#L310
> 
> That looks good, just might be a bit noisy for duplicates, if the payload contains pre-generated IDs ('_id').
> 
> If the client is not generating their own IDs for the Elasticsearch records, then `create` not get 409s, and depending on how efficient Vector ends up being sending `_bulk` requests, it could generate many duplicates in certain cases.
> 
> Does the request code time out on long requests? If so, that will be another problem when the Elasticsearch instance gets over loaded.
> 
> The logic looks like it retries on server errors for any 500 error. We only retry on 500, 503, and 504 with a backoff.
> 
> If the client is using `index`, or using `create` without generating their own IDs, then retries on 5xx errors can overwhelm and Elasticsearch instance.

Awesome feedback @portante, thanks! I'll bother the timber folks to see if they have a preferred handling of this, I know 0.11.x is introducing some automated backoff behaviour which should alleviate some of the overwhelming aspect

File: src/sinks/elasticsearch.rs [hunk 2]
```diff
@@ -94,6 +96,35 @@ impl ElasticSearchAuth {
...
 }
 
+#[derive(Deserialize, Serialize, Clone, Debug)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub enum BulkAction {
+    Index,
+    Create,
+}
+
+impl Default for BulkAction {
```
Comment (jszwedko) [inline src/sinks/elasticsearch.rs:106]: Not a big deal, but I think this could also be done as:

```rust
#[derive(Derivative, Deserialize, Serialize, Clone, Debug)]
#[serde(deny_unknown_fields, rename_all = "snake_case")]
#[derivative(Default)]
pub enum BulkAction {
   #[derivative(Default)]
    Index,
    Create,
}
```

To avoid needing to implement the `Default` trait.

**Review by jszwedko — APPROVED** [review]
I'm actually curious to get @binarylogic's thoughts here given his familiarity with ES.

Right now we don't have any handling of partial bulk insert failures (see https://github.com/timberio/vector/issues/140 for discussion of improving that) so records rejected as duplicates should just be dropped as the request is not retried unless a 500-level error occurs. Assuming the 500 represents a partially processed request, the retried request should just see individual errors for the duplicate records (as I understand it) but still process the rest of the records.

In both the `index` and `create` case there is a chance for there to be duplicate records if `_id` is not set.

**Review by jszwedko — CHANGES_REQUESTED** [review]
Meant to leave that review as a "Comment". We still need to update the docs and, if we could, add a test here (I think just testing `encode_event` is fine rather than a full integration test).

File: src/sinks/elasticsearch.rs (hunk 2 above)
Comment (spencergilbert) [reply src/sinks/elasticsearch.rs:106]: Can do

**Comment by spencergilbert** [comment]
> Meant to leave that review as a "Comment". We still need to update the docs and, if we could, add a test here (I think just testing `encode_event` is fine rather than a full integration test).

Yep - If the implementation is a good enough option as is, I'll get the branch updated, tested, and documented. I'll keep an eye out in case @binarylogic has a different opinion

**Comment by spencergilbert** [comment]
> Meant to leave that review as a "Comment". We still need to update the docs and, if we could, add a test here (I think just testing `encode_event` is fine rather than a full integration test).

I believe the integration tests would need to use a more recent version of Elasticsearch to test datastreams - I can open an issue around upgrading ES - if we're interested in that scenario.

**Comment by jamtur01** [comment]
@spencergilbert Yes please open that issue.

**Comment by jamtur01** [comment]
@spencergilbert Or if you want to include in this PR that'd be fine too.

**Comment by spencergilbert** [comment]
> @spencergilbert Or if you want to include in this PR that'd be fine too.

I'll keep it separate, there might be some breaking changes between 6.6.x (I think this is what integration testing is using) and the introduction of data streams at ~7.9. Besides, _technically_ you can do Create actions regardless of datastream usage.

**Comment by spencergilbert** [comment]
I've added a test (which probably needs tweaking to not use exclude_fields), docs, and improved the default code per @jszwedko's suggestion

**Comment by spencergilbert** [comment]
> Adding a new `bulk_action` option is a great first step, and as @jszwedko pointed out, we will need to improve the documentation so users understand the behavioral differences.

@binarylogic how in detail do we want to go on vector's docs versus linking to the elasticsearch documentation?

**Comment by binarylogic** [comment]
Definitely link out. I just want to point out that the `create` action must be used with data streams. I would add something to the `requirements` attribute about that, as well as a `how_it_works` section with a "Data streams" title.

**Comment by spencergilbert** [comment]
> Definitely link out. I just want to point out that the `create` action must be used with data streams. I would add something to the `requirements` attribute about that, as well as a `how_it_works` section with a "Data streams" title.

Are `requirements` being displayed somewhere today? I checked the Clickhouse sink and WASM transform that have requirements specified in their cue file, but don't see it on the doc site? Wanted to check against existing examples before writing something up.

File: docs/reference/components/sinks/elasticsearch.cue [hunk 3]
```diff
@@ -236,9 +246,19 @@ components: sinks: elasticsearch: {
 			title: "Conflicts"
 			body: """
 				Vector [batches](#buffers--batches) data flushes it to Elasticsearch's
-				[`_bulk` API endpoint][urls.elasticsearch_bulk]. All events are inserted
+				[`_bulk` API endpoint][urls.elasticsearch_bulk]. By default, all events are inserted
 				via the `index` action. In the case of an conflict, such as a document with the
```
Comment (portante) [inline docs/reference/components/sinks/elasticsearch.cue:250]: `s/an/a/`

File: docs/reference/components/sinks/elasticsearch.cue [hunk 4]
```diff
@@ -236,9 +246,19 @@ components: sinks: elasticsearch: {
...
-				same `id`, Vector will add or _replace_ the document as necessary.
+				same `id`, Vector will add or _replace_ the document as necessary. If `bulk_action` is
```
Comment (portante) [inline docs/reference/components/sinks/elasticsearch.cue:251]: Is it Vector that will add or replace or Elasticsearch?

With `index`, if an action entry already has an `_id` field specified, won't Elasticsearch perform an "update" (causing segment merges due to the implicit delete that happens)?  And if an action entry does _not_ have an `_id` field, won't Elasticsearch will generate a unique `_id` value itself, creating a new document in the target index?

So with the `index` action there is really no notion of "conflict".

The notion of a conflict only comes with the `create` action and the action entry has an `_id` field specified.  In that case, if a document exists with the same `_id`, Elasticsearch returns a 409 conflict, else the new document is added.  If the `create` action is used without an `_id` field specified, no conflict will occur because a unique ID will be assigned by Elasticsearch and inserted.

Comment (spencergilbert) [reply docs/reference/components/sinks/elasticsearch.cue:255]: If you specify an `_id` and "index" a document if there is an existing document with that `_id` it should be replaced upon indexing. Copying from the ES docs:

```
index

    (Optional, string) Indexes the specified document. If the document exists, replaces the document and increments the version.
```

I believe the original documentation is accurate if considering the "conflict" being the same as replacing an existing document.

File: docs/reference/components/sinks/elasticsearch.cue (hunk 3 above)
Comment (spencergilbert) [reply docs/reference/components/sinks/elasticsearch.cue]: Thanks, fixed.

File: docs/reference/components/sinks/elasticsearch.cue (hunk 4 above)
Comment (portante) [reply docs/reference/components/sinks/elasticsearch.cue:255]: Hopefully we have somehow communicated our definition of "conflict", which is different from how Elasticsearch uses it (conflict only mentioned with "create").

And it would be great to make sure the wording here places the ownership of the behaviors taken on Elasticsearch and not on Vector (replace "Vector" with "Elasticsearch" in "Vector will add or _replace_ the document as necessary").

Comment (spencergilbert) [reply docs/reference/components/sinks/elasticsearch.cue:255]: @binarylogic let me know your thoughts on this - I had intentionally left the original descriptions "as is" as possible.

**Review by jszwedko — APPROVED** [review]
This looks good to me, thanks @spencergilbert !

I agree with @binarylogic that it shouldn't close https://github.com/timberio/vector/pull/5126 so we can think about any other changes that might improve data streams support, but this should unblock people who want to use it now.

File: docs/reference/components/sinks/elasticsearch.cue [hunk 5]
```diff
@@ -76,7 +76,11 @@ components: sinks: elasticsearch: {
 			"x86_64-unknown-linux-musl":  true
 		}
 
-		requirements: []
+		requirements: [
+			#"""
+				Elasticsearch's Data streams feature require Vector to be configured with the `create` `bulk_action`. *This is not enabled by default.*
```
Comment (jszwedko) [inline docs/reference/components/sinks/elasticsearch.cue:81]:
```suggestion
				Elasticsearch's Data streams feature requires Vector to be configured with the `create` `bulk_action`. *This is not enabled by default.*
```

File: docs/reference/components/sinks/elasticsearch.cue [hunk 6]
```diff
@@ -151,6 +155,16 @@ components: sinks: elasticsearch: {
 				}
 			}
 		}
+		bulk_action: {
+			common:      false
+			description: "Action to use when making requests to the Bulk API. Supports `index` and `create`."
```
Comment (jszwedko) [inline docs/reference/components/sinks/elasticsearch.cue:160]:
```suggestion
			description: "Action to use when making requests to the [Elasticsearch Bulk API](elasticsearch_bulk). Supports `index` and `create`."
```

File: docs/reference/components/sinks/elasticsearch.cue (hunk 4 above)
Comment (jszwedko) [reply docs/reference/components/sinks/elasticsearch.cue:255]: I tend to agree that "conflict" is a bad word for what happens for `index` actions. I'd reword this as:

```
Vector [batches](#buffers--batches) data flushes it to Elasticsearch's [`_bulk` API endpoint][urls.elasticsearch_bulk]. By default, all events are inserted via the `index` action which will update documents if an existing one has the same `id`. If `bulk_action` is configured with `create`, Elasticsearch will _not_ replace an existing document and instead return a conflict error.
```

**Review by jszwedko — COMMENTED** [review]
A couple of documentation notes

**Comment by spencergilbert** [comment]
Thanks @jszwedko I merged all your suggestions for the docs, they definitely look better :smile: how do we look now?

**Review by jszwedko — APPROVED** [review]
Looks good! Thank you.

**Comment by jszwedko** [comment]
Noting that the test failures have been fixed on master.

**Comment by jszwedko** [comment]
Thanks for all of your work on this @spencergilbert !
